### PR TITLE
Add negated predicate functions and selection by type

### PR DESCRIPTION
**Pull request by Yijia-Chen**

### CODE REVIEW

**Comment by Yijia-Chen** [comment]
This `NamedTupleUtilities` module should be complete for a while. The function names should be reviewed to see if they can be more descriptive. Request review from @davidanthoff

File: src/NamedTupleUtilities.jl [hunk 1]
```diff
@@ -51,13 +51,44 @@ julia> range((a=1,b=2,c=3),Val(:a),Val(:b))
...
+"""
+    range(a::NamedTuple, b::Val{n}, c::Val{n})
+Return a NamedTuple which retains `b`th to `c`th fields in `a`. 
+If `b` is greater than or equal to `c`, then it will return the empty NamedTuple.
+```jldoctest
+julia> QueryOperators.NamedTupleUtilities.range((a=1,b=2,c=3),1,2)
+(a = 1, b = 2)
+```
+"""
+@generated function range(a::NamedTuple{an}, b::Int, c::Int) where {an}
```
Comment (davidanthoff) [inline src/NamedTupleUtilities.jl]: I think this needs to be ``@generated function range(a::NamedTuple{an}, ::Val{b}, ::Val{c}) where {an, b:<Int, c<:Int}``? I don't think the current version can work, see my next comment.

File: src/NamedTupleUtilities.jl [hunk 2]
```diff
@@ -51,13 +51,44 @@ julia> range((a=1,b=2,c=3),Val(:a),Val(:b))
...
+    rangeStarted = false
+    names = Symbol[]
+    count = 1
+    for n in an
+        if count == b
```
Comment (davidanthoff) [inline src/NamedTupleUtilities.jl]: With the current version, I believe ``b`` will have the value ``Int``, and comparing that with ``1`` will always be ``false``. The issue is that in a generated function, you *never* have access to the values passed, you always just get the types of the arguments. We can get around that by using the ``Val{b}`` method signature I indicated above.

Also, we should write a unit test, I believe that should have caught the issue here?

File: src/NamedTupleUtilities.jl [hunk 3]
```diff
@@ -91,48 +122,107 @@ ERROR: duplicate field name in NamedTuple: "c" is not unique
...
+end
+
+"""
+    startswith(a::NamedTuple, b::Val{n})
+Return a NamedTuple which retains the fields with names that do not start with `b` in `a`. 
+```jldoctest
+julia> QueryOperators.NamedTupleUtilities.not_startswith((abc=1,bcd=2,cde=3),Val(:a))
+(bcd = 2, cde = 3)
+```
+"""
```
Comment (davidanthoff) [inline src/NamedTupleUtilities.jl]: This doc string seems to get attached to the wrong function?

File: src/NamedTupleUtilities.jl [hunk 4]
```diff
@@ -91,48 +122,107 @@ ERROR: duplicate field name in NamedTuple: "c" is not unique
...
+end
+
+"""
+    endswith(a::NamedTuple, b::Val{n})
+Return a NamedTuple which retains the fields with names that do not end with `b` in `a`. 
+```jldoctest
+julia> QueryOperators.NamedTupleUtilities.not_endswith((abc=1,bcd=2,cde=3),Val(:d))
+(abc = 1, cde = 3)
+```
+"""
```
Comment (davidanthoff) [inline src/NamedTupleUtilities.jl]: Same here, this doc string should go with the previous function, right?

File: src/NamedTupleUtilities.jl [hunk 5]
```diff
@@ -91,48 +122,107 @@ ERROR: duplicate field name in NamedTuple: "c" is not unique
...
+end
+
+"""
+    not_occursin(a::NamedTuple, b::Val{n})
+Return a NamedTuple which retains the fields without names containing `b` as a substring. 
+```jldoctest
+julia> QueryOperators.NamedTupleUtilities.not_occursin((abc=1,bcd=2,cde=3),Val(:d))
+(abc = 1,)
+```
+"""
```
Comment (davidanthoff) [inline src/NamedTupleUtilities.jl]: Same thing here.

File: src/NamedTupleUtilities.jl [hunk 6]
```diff
@@ -91,48 +122,107 @@ ERROR: duplicate field name in NamedTuple: "c" is not unique
...
+
+"""
+    oftype(a::NamedTuple, b::DataType)
+Returns a NamedTuple which retains the fields whose elements have type `b`.
+```jldoctest
+julia> QueryOperators.NamedTupleUtilities.oftype((a = [4,5,6], b = [3.,2.,1.], c = ["He","llo","World!"]), Int64)
+(a = [4, 5, 6],)
+```
+"""
+@generated function oftype(a::NamedTuple{an}, b::DataType) where {an}
```
Comment (davidanthoff) [inline src/NamedTupleUtilities.jl]: Does this actually work? I would have thought we need to do something like ``::Val{b}`` here?

**Comment by davidanthoff** [comment]
Also, please merge ``master`` into this branch to pick up some important bug fixes we have merged in the meantime.

**Comment by Yijia-Chen** [comment]
@davidanthoff Thanks for your review. I just made the requested changes.

**Pull request merged**